business_rules/utils.py

```diff
@@ -85,3 +86,113 @@ def params_dict_to_list(params):
+            ', '.join(invalid_params), method_type_name, method.__name__))
+
+
+def validate_rule_data(variables, actions, rule):
+    """
+    validate_rule_data is used to check a generated rule against a set of variables and actions
+    :param variables:
+    :param actions:
+    :param rule:
+    :return: bool
+    :raises AssertionError:
+    """
+    def validate_root_keys(rule):
+        """
+        Check the root object contains both 'actions' & 'conditions'
+        """
+        root_keys = rule.keys()
+        if 'actions' not in root_keys:
+            raise AssertionError('Missing "{}" key'.format('actions'))
+        if 'conditions' not in root_keys:
+            raise AssertionError('Missing "{}" key'.format('conditions'))
+
+    def validate_condition_operator(condition, rule_schema):
+        """
+        Check provided condition contains a valid operator
+        """
+        for item in rule_schema.get('variables'):
+            if item.get('name') == condition.get('name'):
+                condition_field_type = item.get('field_type')
+                variable_operators = rule_schema.get('variable_type_operators', {}).get(condition_field_type, [])
+                for operators in variable_operators:
+                    if operators['name'] == condition['operator']:
+                        return True
+                raise AssertionError('Unknown operator "{}"'.format(condition['operator']))
+        raise AssertionError('Name "{}" not supported'.format(condition.get('name')))
+
+    def validate_condition_name(condition, variables):
+        """
+        Check provided condition contains a 'name' key and the value is valid
+        """
+        condition_name = condition.get('name')
+        if not condition_name:
+            raise AssertionError('Missing condition "name" key in {}'.format(condition))
+        if not hasattr(variables, condition_name):
+            raise AssertionError('Unknown condition "{}"'.format(condition_name))
+
+    def validate_condition(condition, variables, rule_schema):
+        validate_condition_name(condition, variables)
+        validate_condition_operator(condition, rule_schema)
+        method = getattr(variables, condition.get('name'))
+        params = condition.get('params', {})
+        check_params_valid_for_method(method, params, method_type.METHOD_TYPE_VARIABLE)
+
+    def validate_conditions(input_conditions, rule_schema):
+        """
+        Recursively check all levels of input conditions
+        """
+        if isinstance(input_conditions, list):
+            for condition in input_conditions:
+                validate_conditions(condition, rule_schema)
+        if isinstance(input_conditions, dict):
+            keys = input_conditions.keys()
+            if 'any' in keys or 'all' in keys:
+                if len(keys) > 1:
+                    raise AssertionError('Expected ONE of "any" or "all" but found {}'.format(keys))
```

Comment:
Do you not want the message to specify the quantity of keys? Here you will just return a list of the keys rather than the quantity `len(keys)`